dbms/src/Encryption/FileProvider.cpp

```diff
@@ -50,7 +61,15 @@ WritableFilePtr FileProvider::newWritableFile(
     int flags,
     mode_t mode) const
 {
-    WritableFilePtr file = std::make_shared<PosixWritableFile>(file_path_, truncate_if_exists_, flags, mode, write_limiter_);
+    WritableFilePtr file;
+    if (S3::S3FilenameView::fromKey(file_path_).isValid())
+    {
+        file = S3::S3WritableFile::create(file_path_);
+    }
```

Comment:
I think adding new methods as follows is more explicit than mixing them in the existing methods
* `RandomAccessFilePtr FileProvider::newRandomAccessFile(const String & s3key)`
* `WritableFilePtr FileProvider::newS3WritableFile(const String & s3key)`